eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/ServiceClientInstantiationCheck.java

```diff
@@ -0,0 +1,167 @@
+     */
+    private boolean hasServiceClientAnnotation(DetailAST token) {
+        DetailAST modifiersToken = token.findFirstToken(TokenTypes.MODIFIERS);
+        if (modifiersToken != null) {
+            for (DetailAST ast = modifiersToken.getFirstChild(); ast != null; ast = ast.getNextSibling()) {
+                if (ast.getType() == TokenTypes.ANNOTATION) {
+                    DetailAST annotationIdent = ast.findFirstToken(TokenTypes.IDENT);
+                    if (annotationIdent != null && SERVICE_CLIENT.equals(annotationIdent.getText())) {
+                        return true;
+                    }
+                }
+            }
+        }
+        return false;
+    }
+
+    /**
+     *  Checks for public or protected constructor for the service client class.
+     *  Log error if the service client has public or protected constructor.
+     *
+     * @param token the CTOR_DEF AST node
+     */
+    private void checkNonPublicOrProtectedConstructor(DetailAST token) {
+        DetailAST modifiersToken = token.findFirstToken(TokenTypes.MODIFIERS);
+        AccessModifier accessModifier = CheckUtil.getAccessModifierFromModifiersToken(modifiersToken);
+        if (accessModifier.equals(AccessModifier.PUBLIC) || accessModifier.equals(AccessModifier.PROTECTED)) {
+            log(modifiersToken, "Subclass of ServiceClient should not have any public or protected constructor.");
+        }
+    }
+
+    /**aa
+     * Checks for public static method named 'builder'. Should avoid to use method name, 'builder'.
+     *
+     * @param token the METHOD_DEF AST node
+     */
+    private void checkNonPublicStaticBuilderMethodName(DetailAST token) {
+        DetailAST methodNameToken = token.findFirstToken(TokenTypes.IDENT);
+        if (methodNameToken != null && BUILDER.equals(methodNameToken.getText())) {
+            DetailAST modifiersToken = token.findFirstToken(TokenTypes.MODIFIERS);
+            AccessModifier accessModifier = CheckUtil.getAccessModifierFromModifiersToken(modifiersToken);
+            if (accessModifier.equals(AccessModifier.PUBLIC)
+                && modifiersToken.branchContains(TokenTypes.LITERAL_STATIC)) {
+                log(modifiersToken, "Subclass of ServiceClient should not have a public static method named ''builder''.");
+            }
+        }
+    }
+
+    /**
+     * Checks for the variable field of the subclass of ServiceClient.
+     * These fields should be final because these classes supposed to be immutable class.
+     *
+     * @param token the OBJBLOCK AST node
+     */
+    private void checkClassFieldFinal(DetailAST token) {
+        for (DetailAST ast = token.getFirstChild(); ast != null; ast = ast.getNextSibling()) {
+            if (TokenTypes.VARIABLE_DEF == ast.getType()) {
+                DetailAST modifiersToken = ast.findFirstToken(TokenTypes.MODIFIERS);
+                if (modifiersToken != null && !modifiersToken.branchContains(TokenTypes.FINAL)) {
```

Comment:
What happens if modifiersToken is null? Does that indicate an error scenario or that there are no modifiers (in which case this should be an error)?

Reply:
The 'modifiersToken' will never be null but will have no child node,

Reply:
If it is never null you shouldn't need to null check here?

Reply:
updated to 
```java
//VARIABLE_DEF token will always MODIFIERS token. If there is no modifier at the variable, no child under
            // MODIFIERS token
            if (!modifiersToken.branchContains(TokenTypes.FINAL)) {
                log(modifiersToken, "The variable field of @ServiceClient should be final. These classes supposed to be immutable.");
            }
            }
```